File: packages/contracts/src/domain/smart-playlist.ts
/**
 * Smart (dynamic, rule-based) playlist contracts.
 *
 * A smart playlist persists only a rule definition; its tracks are evaluated
 * dynamically against the library at read time, so it auto-updates as the
 * library changes. The main process translates these rules into a single SQL
 * query (see the smart-playlists IPC handler).
 */

/**
 * Track columns a rule can match against. This tuple is the single source of
 * truth: the `SmartPlaylistField` union, the IPC zod enum, and the renderer's
 * per-field operator map are all derived from it, so adding a field here is the
 * only edit needed to thread it through every layer.
 */
export const SMART_PLAYLIST_FIELDS = [
  'genre',
  'artist',
  'album',
  'title',
  'year',
  'playCount',
  'isFavorite',
  'dateAdded',
  'lastPlayed',
  'bpm',
  'duration',
  'loudnessLufs',
  'musicalKey',
] as const;

/** Track columns a rule can match against. */
export type SmartPlaylistField = (typeof SMART_PLAYLIST_FIELDS)[number];

/**
 * Comparison operators. Applicability depends on the field's type. Source of
 * truth for the `SmartPlaylistOperator` union and the IPC zod enum.
 */
export const SMART_PLAYLIST_OPERATORS = [
  'is',
  'isNot',
  'contains',
  'greaterThan',
  'lessThan',
  'between',
  'inLastDays',
  'notInLastDays',
] as const;

/** Comparison operators. Applicability depends on the field's type. */
export type SmartPlaylistOperator = (typeof SMART_PLAYLIST_OPERATORS)[number];

/**
 * A single rule. `value` / `valueTo` semantics depend on the operator:
 * - `between` uses both `value` (lower) and `valueTo` (upper).
 * - `inLastDays` / `notInLastDays` use `value` as a day count, against
 *   `dateAdded` or `lastPlayed`.
 * - `isFavorite` uses `value` as a boolean ('true'/'false').
 * - everything else compares `value` against the field.
 *
 * # How NULL behaves
 *
 * `bpm`, `duration`, `loudnessLufs`, `year` and `musicalKey` are all nullable —
 * for the analysis columns, NULL means "not analysed yet" (or "analysed and
 * nothing detectable"; the schema collapses the two). SQL three-valued logic
 * decides what that means and every evaluator inherits it unchanged: a NULL
 * operand satisfies **no** comparison, `isNot` included. An unanalysed track is
 * therefore excluded by "bpm is not 120" just as it is by "bpm is 120". This is
 * the pre-existing behaviour of `year`, kept rather than special-cased: the
 * alternative — treating "unknown" as "does not equal" — makes an `all`
 * definition silently fill up with unanalysed tracks.
 *
 * `lastPlayed` is the deliberate exception, because its NULL is *meaningful*
 * rather than missing: a track with no play history has never been played, so
 * it satisfies `notInLastDays` for any day count. See the evaluators.
 */
export interface SmartPlaylistRule {
  field: SmartPlaylistField;
  operator: SmartPlaylistOperator;
  value: string;
  valueTo?: string;
}

/** How multiple rules combine. */
export type SmartPlaylistMatchType = 'all' | 'any';

/** Sort directions an `orderBy` clause can take. */
export const SMART_PLAYLIST_SORT_DIRECTIONS = ['asc', 'desc'] as const;

/** Sort direction for an `orderBy` clause. */
export type SmartPlaylistSortDirection = (typeof SMART_PLAYLIST_SORT_DIRECTIONS)[number];

/**
 * An explicit sort for a definition's results, replacing the default library
 * order (newest first). Paired with `limit` this is what makes "top 25 most
 * played" and "50 least recently played" expressible.
 *
 * Ordering by `lastPlayed` sorts on the most recent play, and never-played
 * tracks sort as NULL — lowest, so they come *first* ascending. That is the
 * semantically right end: a track never played is the least recently played
 * thing in the library.
 */
export interface SmartPlaylistOrderBy {
  field: SmartPlaylistField;
  direction: SmartPlaylistSortDirection;
}

/**
 * The persisted rule definition (stored JSON-serialized in `rules`).
 *
 * # Storage shape, and why there is no migration
 *
 * The `rules` column has always held a JSON *array* of rules, and rows written
 * by builds older than `limit`/`orderBy` still do. Rather than add columns —
 * which the frozen v1 migration ledger makes expensive — the column now accepts
 * two shapes and readers accept both:
 *
 * - a bare array, `[...]` — the legacy shape, no limit and no sort;
 * - an envelope, `{ "rules": [...], "limit": 25, "orderBy": {...} }`.
 *
 * Writers emit the envelope **only** when a limit or a sort is actually set, so
 * a definition that uses neither round-trips byte-identically to what a v1
 * build would have written and stays readable by one.
 */
export interface SmartPlaylistDefinition {
  matchType: SmartPlaylistMatchType;
  rules: SmartPlaylistRule[];
  /** Maximum tracks to return. Omitted or non-positive means unbounded. */
  limit?: number;
  /** Explicit sort, replacing the default library order. */
  orderBy?: SmartPlaylistOrderBy;
}

/** A persisted smart playlist row (rules parsed back into structured form). */
export interface SmartPlaylist {
  id: string;
  name: string;
  description: string | null;
  matchType: SmartPlaylistMatchType;
  rules: SmartPlaylistRule[];
  limit?: number;
  orderBy?: SmartPlaylistOrderBy;
  createdAt: string;
  updatedAt: string;
}

File: crates/shiranami-core/src/models/smart_playlist.rs
//! Smart (rule-based) playlists, ported from
//! `packages/contracts/src/domain/smart-playlist.ts`.
//!
//! A smart playlist persists only a rule definition; its tracks are evaluated
//! against the library at read time, so it auto-updates as the library changes.
//! Phase 7 translates [`SmartPlaylistDefinition`] into a single SQL query.

use serde::{Deserialize, Serialize};
use specta::Type;

/// Track columns a rule can match against.
///
/// The TypeScript source keeps this as one `SMART_PLAYLIST_FIELDS` tuple from
/// which the union, the zod enum and the renderer's per-field operator map are
/// all derived. The Rust enum is that single source of truth.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub enum SmartPlaylistField {
    /// `tracks.genre`.
    Genre,
    /// `tracks.artist`.
    Artist,
    /// `tracks.album`.
    Album,
    /// `tracks.title`.
    Title,
    /// `tracks.year`.
    Year,
    /// `tracks.play_count`.
    PlayCount,
    /// `tracks.is_favorite`.
    IsFavorite,
    /// `tracks.created_at`.
    DateAdded,
    /// The most recent library play, from `play_history` — no column of its
    /// own. `NULL` (never played) is meaningful here rather than missing.
    LastPlayed,
    /// `tracks.bpm`. `NULL` until the analysis engine has run.
    Bpm,
    /// `tracks.duration`, in seconds.
    Duration,
    /// `tracks.loudness_lufs`. `NULL` until the analysis engine has run.
    LoudnessLufs,
    /// `tracks.musical_key`, the stored Camelot/name string.
    MusicalKey,
}

/// Comparison operators. Applicability depends on the field's type.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub enum SmartPlaylistOperator {
    /// Exact equality.
    Is,
    /// Exact inequality.
    IsNot,
    /// Substring match (`LIKE`, with the `ESCAPE '\'` guard).
    Contains,
    /// Strictly greater than.
    GreaterThan,
    /// Strictly less than.
    LessThan,
    /// Inclusive range across `value` (lower) and `value_to` (upper).
    Between,
    /// `date_added` / `last_played` within the last `value` days.
    InLastDays,
    /// The negation of [`Self::InLastDays`].
    ///
    /// For `last_played` this deliberately includes tracks with no play history
    /// at all: never played satisfies "not played in the last N days" for every
    /// N. That is the rule people actually want, and the one a
    /// `MAX(played_at) < cutoff` comparison quietly gets wrong.
    NotInLastDays,
}

/// How multiple rules combine.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Type)]
#[serde(rename_all = "lowercase")]
pub enum SmartPlaylistMatchType {
    /// Every rule must match (SQL `AND`).
    All,
    /// Any rule may match (SQL `OR`).
    Any,
}

/// A single rule.
///
/// `value` / `value_to` semantics depend on the operator:
/// - [`SmartPlaylistOperator::Between`] uses both, as lower and upper bound.
/// - [`SmartPlaylistOperator::InLastDays`] reads `value` as a day count.
/// - [`SmartPlaylistField::IsFavorite`] reads `value` as `"true"` / `"false"`.
/// - everything else compares `value` against the field.
///
/// The value stays a `String` on purpose — it is a heterogeneously-typed literal
/// the renderer builds from a text input, and Phase 7 parses it per operator.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct SmartPlaylistRule {
    /// Column to match against.
    pub field: SmartPlaylistField,
    /// Comparison to apply.
    pub operator: SmartPlaylistOperator,
    /// Primary operand.
    pub value: String,
    /// Upper bound, for [`SmartPlaylistOperator::Between`] only.
    #[specta(optional)]
    pub value_to: Option<String>,
}

/// Sort direction for a [`SmartPlaylistOrderBy`].
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Type)]
#[serde(rename_all = "lowercase")]
pub enum SmartPlaylistSortDirection {
    /// Ascending. SQLite sorts `NULL` lowest, so never-played sorts first.
    Asc,
    /// Descending.
    Desc,
}

/// An explicit sort, replacing the default library order (newest first).
///
/// Paired with [`SmartPlaylistDefinition::limit`] this is what makes "top 25
/// most played" and "50 least recently played" expressible.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct SmartPlaylistOrderBy {
    /// The field to sort on.
    pub field: SmartPlaylistField,
    /// Which way.
    pub direction: SmartPlaylistSortDirection,
}

/// The persisted rule definition, stored JSON-serialized in the `rules` column.
///
/// # Storage shape
///
/// The column has always held a JSON *array* of rules, and rows written before
/// `limit`/`order_by` existed still do. Rather than migrate — v1's ledger is
/// frozen by [`crate`]'s adoption contract — the column accepts two shapes and
/// readers take both: a bare array (legacy, no limit and no sort), or an
/// envelope `{"rules": [...], "limit": 25, "orderBy": {...}}`. Writers emit the
/// envelope only when a limit or a sort is set, so a definition using neither
/// round-trips exactly as an older build would have written it.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct SmartPlaylistDefinition {
    /// How the rules combine.
    pub match_type: SmartPlaylistMatchType,
    /// The rules themselves. An empty list matches the whole library.
    pub rules: Vec<SmartPlaylistRule>,
    /// Maximum tracks to return. `None` means unbounded.
    #[specta(optional)]
    pub limit: Option<u32>,
    /// Explicit sort, replacing the default library order.
    #[specta(optional)]
    pub order_by: Option<SmartPlaylistOrderBy>,
}

/// A persisted smart playlist row, with `rules` parsed back into structured form.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct SmartPlaylist {
    /// Primary key (UUID v4).
    pub id: String,
    /// Display name.
    pub name: String,
    /// Free-text description; nullable in SQLite, and nullable on the wire.
    pub description: Option<String>,
    /// How the rules combine.
    pub match_type: SmartPlaylistMatchType,
    /// The rules themselves.
    pub rules: Vec<SmartPlaylistRule>,
    /// Maximum tracks to return. `None` means unbounded.
    #[specta(optional)]
    pub limit: Option<u32>,
    /// Explicit sort, replacing the default library order.
    #[specta(optional)]
    pub order_by: Option<SmartPlaylistOrderBy>,
    /// ISO-8601 creation timestamp.
    pub created_at: String,
    /// ISO-8601 last-update timestamp.
    pub updated_at: String,
}
